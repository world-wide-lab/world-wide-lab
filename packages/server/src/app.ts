import path from "path";
import cors from "cors";
import express, { Request, Response } from "express";
import helmet from "helmet";

import { admin, adminRouter } from "./admin";
import api from "./api";
import apiDocs from "./api-docs";
import { routerProtectedWithoutAuthentication } from "./api/protected";
import config from "./config";
import { logger } from "./logger";

const app = express();

// app.use(helmet());
app.use(cors());
app.use(express.json());

// Log all HTTP requests
if (config.logging.http) {
  logger.info("Logging HTTP requests");

  app.use((req, res, next) => {
    logger.http(`${req.method} ${req.url} ${JSON.stringify(req.body)}`, {
      request: {
        query: req.query,
        params: req.params,
        body: req.body,
        headers: req.headers,
      },
    });
    next();
  });
}

app.use("/static", express.static(path.join(__dirname, "..", "static")));

app.get("/", async (req: Request, res: Response) => {
  res.send(`
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>World-Wide-Lab</title>
    </head>
    <body>
      World-Wide-Lab is running 🌐🧑‍🔬👩‍🔬👨‍🔬 <br />
      <ul>
        <li>The latest API can be found under <a href="/v1/">/v1/</a></li>
      </ul>
    </body>
  </html>
  `);
});

// Mount the API router under v1
app.use("/v1", api);

// Mount the api-docs
if (config.apiDocs.enabled) {
  app.use("/api-docs", apiDocs);
}

// Use adminJS
if (config.admin.enabled) {
  // Make protected API routes available in Admin UI
  adminRouter.use("/wwl/", routerProtectedWithoutAuthentication);

  app.use(admin.options.rootPath, adminRouter);
}

export default app;
