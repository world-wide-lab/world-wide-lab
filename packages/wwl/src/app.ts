import path from 'path'
import express, {Request, Response} from 'express';
import helmet from 'helmet';
import cors from 'cors';

import api from './api';
import apiDocs from './api-docs';
import { admin, adminRouter } from './admin'
import config from './config'

const app = express();

// app.use(helmet());
app.use(cors());
app.use(express.json());

app.use('/static', express.static(path.join(__dirname, '..', 'static')))

app.get("/", async (req: Request, res: Response) => {
  res.redirect("/v1/");
});

// Mount the API router under v1
app.use('/v1', api);

// Mount the api-docs
if (config.apiDocs.enabled) {
  app.use('/api-docs', apiDocs);
}

// Use adminJS
if (config.admin.enabled) {
  app.use(admin.options.rootPath, adminRouter)
}

export default app;
