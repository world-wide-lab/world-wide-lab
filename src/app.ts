import 'dotenv/config';
import express, {Request, Response} from 'express';
import helmet from 'helmet';
import cors from 'cors';

import api from './api';

const app = express();

app.use(helmet());
app.use(cors());
app.use(express.json());

app.get("/", async (req: Request, res: Response) => {
  res.redirect("/v1/");
});

// Mount the API router under v1
app.use('/v1', api);

export default app;
