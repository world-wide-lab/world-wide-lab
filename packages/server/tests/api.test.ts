// Set up fake environment variables
import "./setup_env";

import request from "supertest";
import app from "../src/app";
import sequelize from "../src/db";

const STUDY_ID = "abc123";
const API_KEY = process.env.DEFAULT_API_KEY;

const NON_EXISTENT_UUID = "00000000-0000-0000-0000-000000000000";

const endpoint = request(app);

describe("API Routes", () => {
  beforeAll(async () => {
    // Initialize Database
    await sequelize.sync();
  });

  // afterAll(async () => {
  //   // TODO: Clean up any test data
  // });

  let participantId: string;
  let sessionId: string;
  const studyId: string = STUDY_ID;

  describe("POST /participant", () => {
    it("should create a new participant", async () => {
      const response = await endpoint.post("/v1/participant").send();

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("participantId");

      // Set the shared participantId
      participantId = response.body.participantId;
    });
  });

  describe("PUT /participant/:participantId", () => {
    it("should update an existing participant", async () => {
      const response = await endpoint
        .put("/v1/participant/" + participantId)
        .send({
          privateInfo: { lorem: "ipsum" },
          publicInfo: { participantHasDoneSomething: true },
        });

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      const user = await sequelize.models.Participant.findOne({
        where: { participantId },
      });
      expect(user).toHaveProperty("privateInfo", { lorem: "ipsum" });
      expect(user).toHaveProperty("publicInfo", {
        participantHasDoneSomething: true,
      });
    });

    it("should fail when the participant does not exist", async () => {
      const response = await endpoint
        .put("/v1/participant/" + NON_EXISTENT_UUID)
        .send({ privateInfo: { lorem: "ipsum" } });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the participant is invalid", async () => {
      const response = await endpoint
        .put("/v1/participant/" + "some-non-existing-ID")
        .send({ privateInfo: { lorem: "ipsum" } });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("GET /participant/:participantId", () => {
    it("should retrieve public information about a participant", async () => {
      const response = await endpoint
        .get("/v1/participant/" + participantId)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.participantId).toBe(participantId);
      delete response.body.participantId;
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the participant does not exist", async () => {
      const response = await endpoint
        .get("/v1/participant/" + NON_EXISTENT_UUID)
        .send();

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the participant is invalid", async () => {
      const response = await endpoint
        .get("/v1/participant/" + "non-existent-participant-id")
        .send();

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("POST /study", () => {
    it("should create a new study", async () => {
      const response = await endpoint
        .post("/v1/study")
        .send({ studyId: STUDY_ID });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("studyId");
    });

    it("should create a new study with extra info", async () => {
      const response = await endpoint.post("/v1/study").send({
        studyId: STUDY_ID + "_extra",
        privateInfo: { integer: 10 },
        publicInfo: { string: "lorem" },
      });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("studyId");
    });

    it("should validate extra info", async () => {
      const response = await endpoint
        .post("/v1/study")
        .send({ studyId: STUDY_ID + "_extra", privateInfo: 10 });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("POST /session", () => {
    it("should start a new session", async () => {
      const response = await endpoint
        .post("/v1/session")
        .send({ participantId, studyId });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("participantId", participantId);
      expect(response.body).toHaveProperty("studyId", studyId);

      // Set the shared sessionId
      sessionId = response.body.sessionId;
    });

    it("should start a new session with extra info", async () => {
      const response = await endpoint.post("/v1/session").send({
        participantId,
        studyId,
        privateInfo: { integer: 10 },
        publicInfo: { string: "lorem" },
        clientMetadata: { version: "1.0", queryParameters: { test: "test" } },
      });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("participantId", participantId);
      expect(response.body).toHaveProperty("studyId", studyId);
      expect(response.body).toHaveProperty("sessionId");

      const session = await sequelize.models.Session.findOne({
        where: { sessionId: response.body.sessionId },
      });
      // @ts-ignore
      expect(session.metadata).toMatchSnapshot();
    });

    it("missing studyId should lead to an error", async () => {
      const response = await endpoint
        .post("/v1/session")
        .send({ participantId });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
    it("missing participantId should be ok", async () => {
      const response = await endpoint.post("/v1/session").send({ studyId });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("studyId", studyId);
      expect(response.body).toHaveProperty("sessionId");
    });

    it("should validate extra info", async () => {
      const response = await endpoint
        .post("/v1/session")
        .send({ participantId, studyId, privateInfo: 10 });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("POST /session/finish", () => {
    it("should mark a session as finished", async () => {
      const response = await endpoint
        .post("/v1/session/finish")
        .send({ sessionId: sessionId });

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      const session = await sequelize.models.Session.findOne({
        where: { sessionId },
      });
      expect(session).toHaveProperty("finished", true);
    });

    it("should fail when the session does not exist", async () => {
      const response = await endpoint
        .post("/v1/session/finish")
        .send({ sessionId: NON_EXISTENT_UUID });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the sessionId is invalid", async () => {
      const response = await endpoint
        .post("/v1/session/finish")
        .send({ sessionId: "non-existent" });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("PUT /session/:sessionId", () => {
    it("should update a session", async () => {
      const response = await endpoint.put("/v1/session/" + sessionId).send({
        privateInfo: { lorem: "ipsum" },
        publicInfo: { dolor: "sit" },
      });

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      const session = await sequelize.models.Session.findOne({
        where: { sessionId },
      });
      expect(session).toHaveProperty("privateInfo", { lorem: "ipsum" });
      expect(session).toHaveProperty("publicInfo", { dolor: "sit" });
    });

    it("should fail when the session does not exist", async () => {
      const response = await endpoint
        .put("/v1/session/" + NON_EXISTENT_UUID)
        .send({ privateInfo: { lorem: "ipsum" } });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the sessionId is invalid", async () => {
      const response = await endpoint
        .put("/v1/session/" + "non-existent-session-id")
        .send({ privateInfo: { lorem: "ipsum" } });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("GET /session/:sessionId", () => {
    it("should retrieve public information about a session", async () => {
      const response = await endpoint.get("/v1/session/" + sessionId).send();

      expect(response.status).toBe(200);
      expect(response.body.sessionId).toBe(sessionId);
      delete response.body.sessionId;
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the session does not exist", async () => {
      const response = await endpoint
        .get("/v1/session/" + NON_EXISTENT_UUID)
        .send();

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the sessionId is invalid", async () => {
      const response = await endpoint
        .get("/v1/session/" + "non-existent-session-id")
        .send();

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("POST /response", () => {
    it("should submit a response", async () => {
      const response = await endpoint.post("/v1/response").send({
        sessionId,
        name: "test_trail",
        payload: {
          key_1: "value 1",
          key_2: "value 2",
        },
      });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty("responseId");
      const { responseId } = response.body;

      const sessionResponse = await sequelize.models.Response.findOne({
        where: { responseId },
      });
      expect(sessionResponse).toHaveProperty("sessionId", sessionId);
      expect(sessionResponse).toHaveProperty("name", "test_trail");
      expect(sessionResponse).toHaveProperty("payload", {
        key_1: "value 1",
        key_2: "value 2",
      });
    });

    it("should fail to submit a response when the session does not exist", async () => {
      const response = await endpoint.post("/v1/response").send({
        sessionId: NON_EXISTENT_UUID,
        name: "test_trail",
        payload: {
          key_1: "value 1",
          key_2: "value 2",
        },
      });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail to submit a response when the sessionId is invalid", async () => {
      const response = await endpoint.post("/v1/response").send({
        sessionId: "non-existent-session-id",
        name: "test_trail",
        payload: {
          key_1: "value 1",
          key_2: "value 2",
        },
      });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail to submit non-JSON payload", async () => {
      const response = await endpoint.post("/v1/response").send({
        sessionId,
        name: "test_trail",
        payload: "this-is-not-json",
      });

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });

    // These extra responses are for the GET /study/* tests cases
    it("should store additional data", async () => {
      const exampleData = {
        name: "test_trail",
        payload: {
          key_1: "value 1",
          key_2: "value 2",
        },
      };
      // Two more responses for first session
      await endpoint.post("/v1/response").send({ sessionId, ...exampleData });
      await endpoint.post("/v1/response").send({ sessionId, ...exampleData });
      // And one additional responses in a new session
      const newSessionResponse = await endpoint
        .post("/v1/session")
        .send({ participantId, studyId });
      await endpoint
        .post("/v1/response")
        .send({ sessionId: newSessionResponse.body.sessionId, ...exampleData });
    });
  });

  describe("GET /study/:studyId/count/:countType", () => {
    it("should return the correct count (for all sessions)", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/count/all`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.count).toBe(4);
    });

    it("should return the correct count (for only finished sessions)", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/count/finished`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.count).toBe(1);
    });

    it("should fail when the countType does not exist", async () => {
      // Don't pass on console.error message, as it is expected
      jest.spyOn(console, "error").mockImplementation(() => {});

      const response = await endpoint
        .get(`/v1/study/${studyId}/count/non-existent-type`)
        .send();

      expect(response.status).toBe(500);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the study does not exist", async () => {
      const response = await endpoint
        .get(`/v1/study/non-existent-study/count/all`)
        .send();

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("GET /study/:studyId/data/:dataType/json", () => {
    it("should download a raw list of responses", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/responses-raw/json`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(4);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should download a raw list of sessions", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/sessions-raw/json`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(4);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should download a raw list of participant", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/participants-raw/json`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(3);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should download an extracted list of responses", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/responses-extracted-payload/json`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(4);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
      expect(response.body[0].key_1).toBe("value 1");
      expect(response.body[3].key_2).toBe("value 2");
    });

    it("should download an extracted list of responses (in CSV format)", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/responses-extracted-payload/csv`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      const lines = response.text.split(/\r\n|\r|\n/);
      expect(lines.length).toBe(4 + 1);
      expect(lines[0]).toMatchSnapshot();
    });

    it("should handle empty studies as well", async () => {
      const studyIdEmpty = "empty-study";
      // Create new empty study
      await endpoint.post("/v1/study").send({ studyId: studyIdEmpty });

      const response = await endpoint
        .get(`/v1/study/${studyIdEmpty}/data/responses-extracted-payload/json`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(0);
    });

    it("should handle empty studies as well (in CSV format)", async () => {
      const studyIdEmpty = "empty-study-2";
      // Create new empty study
      await endpoint.post("/v1/study").send({ studyId: studyIdEmpty });

      const response = await endpoint
        .get(`/v1/study/${studyIdEmpty}/data/responses-extracted-payload/csv`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.text).toBe("");
    });

    it("should handle studies without payload as well", async () => {
      const studyIdNoPayload = "no-payload-study";
      // Create new empty study and add one session with one response
      await endpoint.post("/v1/study").send({ studyId: studyIdNoPayload });
      const sessionResponse = await endpoint
        .post("/v1/session")
        .send({ participantId, studyId: studyIdNoPayload });
      await endpoint
        .post("/v1/response")
        .send({ sessionId: sessionResponse.body.sessionId });

      const response = await endpoint
        .get(
          `/v1/study/${studyIdNoPayload}/data/responses-extracted-payload/json`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(1);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should require authentication", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/participants-raw/json`)
        .send();

      expect(response.status).toBe(401);
      expect(response.body).toMatchSnapshot();
    });

    it("should require the correct API KEY", async () => {
      const response = await endpoint
        .get(`/v1/study/${studyId}/data/participants-raw/json`)
        .set("Authorization", `Bearer wrong-key`)
        .send();

      expect(response.status).toBe(401);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when the study does not exist", async () => {
      const response = await endpoint
        .get(`/v1/study/non-existent-study/data/participants-raw/json`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(400);
      expect(response.body).toMatchSnapshot();
    });
  });
});
