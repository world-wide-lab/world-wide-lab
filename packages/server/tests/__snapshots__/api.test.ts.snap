// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Routes GET /participant/:participantId should fail when the participant does not exist 1`] = `
{
  "error": "Unknown participantId",
}
`;

exports[`API Routes GET /participant/:participantId should fail when the participant is invalid 1`] = `
{
  "error": "participantId must be a valid UUID",
}
`;

exports[`API Routes GET /participant/:participantId should retrieve public information about a participant 1`] = `
{
  "publicInfo": {
    "participantHasDoneSomething": true,
  },
}
`;

exports[`API Routes GET /session/:sessionId should fail when the session does not exist 1`] = `
{
  "error": "Unknown sessionId",
}
`;

exports[`API Routes GET /session/:sessionId should fail when the sessionId is invalid 1`] = `
{
  "error": "sessionId must be a valid UUID",
}
`;

exports[`API Routes GET /session/:sessionId should retrieve public information about a session 1`] = `
{
  "publicInfo": {
    "dolor": "sit",
  },
}
`;

exports[`API Routes GET /study/:studyId/count/:countType should fail when the countType does not exist 1`] = `
{
  "error": "Failed to retrieve study count",
}
`;

exports[`API Routes GET /study/:studyId/count/:countType should fail when the study does not exist 1`] = `
{
  "error": "Unknown studyId",
}
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should download a raw list of participant 1`] = `
[
  "participantId",
  "createdAt",
  "updatedAt",
  "privateInfo",
  "publicInfo",
  "Sessions.sessionId",
]
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should download a raw list of responses 1`] = `
[
  "responseId",
  "createdAt",
  "updatedAt",
  "name",
  "payload",
  "sessionId",
  "Session.participantId",
]
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should download a raw list of sessions 1`] = `
[
  "sessionId",
  "createdAt",
  "updatedAt",
  "privateInfo",
  "publicInfo",
  "finished",
  "participantId",
  "studyId",
  "metadata",
]
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should download an extracted list of responses (in CSV format) 1`] = `"responseId,createdAt,updatedAt,name,sessionId,key_1,key_2"`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should download an extracted list of responses 1`] = `
[
  "responseId",
  "createdAt",
  "updatedAt",
  "name",
  "sessionId",
  "key_1",
  "key_2",
]
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should fail when the study does not exist 1`] = `
{
  "error": "Unknown studyId",
}
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should handle studies without payload as well 1`] = `
[
  "responseId",
  "createdAt",
  "updatedAt",
  "name",
  "sessionId",
]
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should require authentication 1`] = `
{
  "error": "Authentication via API Key required",
}
`;

exports[`API Routes GET /study/:studyId/data/:dataType/json should require the correct API KEY 1`] = `
{
  "error": "The provided API Key is invalid.",
}
`;

exports[`API Routes POST /response should fail to submit a response when the session does not exist 1`] = `
{
  "error": "Unknown sessionId",
}
`;

exports[`API Routes POST /response should fail to submit a response when the sessionId is invalid 1`] = `
{
  "error": "sessionId must be a valid UUID",
}
`;

exports[`API Routes POST /response should fail to submit non-JSON payload 1`] = `
{
  "error": "payload must be a \`object\` type, but the final value was: \`"this-is-not-json"\`.",
}
`;

exports[`API Routes POST /session missing studyId should lead to an error 1`] = `
{
  "error": "studyId is a required field",
}
`;

exports[`API Routes POST /session should validate extra info 1`] = `
{
  "error": "privateInfo must be a \`object\` type, but the final value was: \`10\`.",
}
`;

exports[`API Routes POST /session/finish should fail when the session does not exist 1`] = `
{
  "error": "Unknown sessionId",
}
`;

exports[`API Routes POST /session/finish should fail when the sessionId is invalid 1`] = `
{
  "error": "sessionId must be a valid UUID",
}
`;

exports[`API Routes POST /session/finish should mark a session as finished 1`] = `
{
  "success": true,
}
`;

exports[`API Routes POST /study should validate extra info 1`] = `
{
  "error": "privateInfo must be a \`object\` type, but the final value was: \`10\`.",
}
`;

exports[`API Routes PUT /participant/:participantId should fail when the participant does not exist 1`] = `
{
  "error": "Unknown participantId",
}
`;

exports[`API Routes PUT /participant/:participantId should fail when the participant is invalid 1`] = `
{
  "error": "participantId must be a valid UUID",
}
`;

exports[`API Routes PUT /participant/:participantId should update an existing participant 1`] = `
{
  "success": true,
}
`;

exports[`API Routes PUT /session/:sessionId should fail when the session does not exist 1`] = `
{
  "error": "Unknown sessionId",
}
`;

exports[`API Routes PUT /session/:sessionId should fail when the sessionId is invalid 1`] = `
{
  "error": "sessionId must be a valid UUID",
}
`;

exports[`API Routes PUT /session/:sessionId should update a session 1`] = `
{
  "success": true,
}
`;
